prov/shm/src/smr_ep.c

```diff
@@ -497,92 +497,118 @@ static int smr_format_mmap(struct smr_ep *ep, struct smr_cmd *cmd,
 		if (cmd->msg.hdr.op == ofi_op_read_req)
 			resp->status = FI_SUCCESS;
-		*next = 0;
+		next_sar_buf++;
 	}
+
+	resp->sar_status = SMR_SAR_READY;
+
 	return *bytes_done - start;
 }
 
-size_t smr_copy_from_sar(struct smr_sar_msg *sar_msg, struct smr_resp *resp,
+size_t smr_copy_from_sar(struct smr_sar_pool *sar_pool, struct smr_resp *resp,
 			 struct smr_cmd *cmd, enum fi_hmem_iface iface,
 			 uint64_t device, const struct iovec *iov, size_t count,
 			 size_t *bytes_done, int *next)
 {
+	struct smr_sar_msg *next_sar_msg;
 	size_t start = *bytes_done;
+	int next_sar_buf = 0;
+	if (resp->sar_status != SMR_SAR_READY)
+		return 0;
 
-	if (sar_msg->sar[0].status == SMR_SAR_READY && !*next) {
-		*bytes_done += ofi_copy_to_hmem_iov(iface, device, iov, count,
-					*bytes_done, sar_msg->sar[0].buf,
-					SMR_SAR_SIZE);
-		sar_msg->sar[0].status = SMR_SAR_FREE;
-		if (cmd->msg.hdr.op != ofi_op_read_req)
-			resp->status = FI_SUCCESS;
-		*next = 1;
-	}
+	while ((*bytes_done < cmd->msg.hdr.size) && (next_sar_buf < cmd->msg.data.buf_batch_size)) {
+		next_sar_msg = smr_freestack_get_entry_from_index(
+		    sar_pool, cmd->msg.data.sar[next_sar_buf]);
+
+		*bytes_done += ofi_copy_to_hmem_iov(
+			iface, device, iov, count, *bytes_done,
+			next_sar_msg->sar.buf, SMR_SAR_SIZE);
 
-	if (*bytes_done < cmd->msg.hdr.size &&
-	    sar_msg->sar[1].status == SMR_SAR_READY && *next) {
-		*bytes_done += ofi_copy_to_hmem_iov(iface, device, iov, count,
-					*bytes_done, sar_msg->sar[1].buf,
-					SMR_SAR_SIZE);
-		sar_msg->sar[1].status = SMR_SAR_FREE;
 		if (cmd->msg.hdr.op != ofi_op_read_req)
 			resp->status = FI_SUCCESS;
-		*next = 0;
+		next_sar_buf++;
 	}
+	resp->sar_status = SMR_SAR_FREE;
 	return *bytes_done - start;
 }
 
-int smr_format_sar(struct smr_cmd *cmd, enum fi_hmem_iface iface, uint64_t device,
-		   const struct iovec *iov, size_t count,
-		   size_t total_len, struct smr_region *smr,
-		   struct smr_region *peer_smr, int64_t id,
-		   struct smr_tx_entry *pending, struct smr_resp *resp)
+static int smr_format_sar(struct smr_ep *ep, struct smr_cmd *cmd,
+		   enum fi_hmem_iface iface, uint64_t device,
+		   const struct iovec *iov, size_t count, size_t total_len,
+		   struct smr_region *smr, struct smr_region *peer_smr,
+		   int64_t id, struct smr_tx_entry *pending,
+		   struct smr_resp *resp)
 {
-	struct smr_sar_msg *sar_msg;
+	int i;
+	uint32_t available_sar_buffers;
 
 	if (!peer_smr->sar_cnt)
 		return -FI_EAGAIN;
 
-	sar_msg = smr_freestack_pop(smr_sar_pool(peer_smr));
+	available_sar_buffers = peer_smr->max_sar_buff_per_peer -
+				smr_peer_data(smr)[id].sar_buff_inuse;
+
+	if (available_sar_buffers == 0)
+		return -FI_EAGAIN;
+
+	if (available_sar_buffers > SMR_BUF_BATCH_MAX)
+		available_sar_buffers = SMR_BUF_BATCH_MAX;
+
+	cmd->msg.data.buf_batch_size =
+	    (((total_len + SMR_SAR_SIZE - 1) / SMR_SAR_SIZE) <
+		     available_sar_buffers
+		 ? ((total_len + SMR_SAR_SIZE - 1) / SMR_SAR_SIZE)
+		 : available_sar_buffers);
```

Comment:
The above flow from the if statement on might be easier to read as:
```
static int smr_sar_avail(smr_region *smr)
{
    return max_sar - sar_inuse;
}

static int smr_sar_needed(size_t len)
{
    // note there may be an existing macro for this -- maybe pages needed?
    return (len + SAR_SIZE - 1) / SAR_SIZE;
}

batch_size = min(MATCH_BATCH, min(smr_sar_avail(..), smr_sar_needed(...))
```

Reply:
We will restructure the code to be more readable along the line you suggested.